src/odemis/driver/npmc.py

```diff
@@ -0,0 +1,845 @@
+                logging.debug("Cancelling failed")
+            return future._was_stopped
+
+    @isasync
+    def reference(self, axes):
+        model.Actuator.reference(self, axes)
+
+    def stop(self, axes=None):
+        self._executor.cancel()
+        # For safety, just force stop every axis
+        for an, aid in self._axis_map.items():
+            if axes is None or an in axes:
+                self.StopMotion(aid)
+        
+    def _createMoveFuture(self):
+        """
+        Return (CancellableFuture): a future that can be used to manage a move
+        """
+        f = CancellableFuture()
+        f._moving_lock = threading.Lock()  # taken while moving
+        f._must_stop = threading.Event()  # cancel of the current future requested
+        f._was_stopped = False  # if cancel was successful
+        f.task_canceller = self._cancelCurrentMove
+        return f
+
+
+class ESPSimulator(object):
+    """
+    Simulates a Newport ESP301
+    Same interface as the serial port
+    """
+
+    def __init__(self, timeout=1):
+        # we don't care about the actual parameters but timeout
+        self.timeout = timeout
+        self._output_buf = ""  # what the commands sends back to the "host computer"
+        self._input_buf = ""  # what we receive from the "host computer"
+
+        self._position = [0, 0, 0]
+        self._start_pos = [0, 0, 0]
+        self._range = [-500, 500]
+        self._target = [0, 0, 0]
+        self._speed = [50, 50, 50]
+        self._accel = [20, 20, 20]
+        self._decel = [20, 20, 20]
+        self._error_stack = []
+        
+        self._time_start = time.time()
+        self._current_move_finish = time.time()
+        self._moving = False
+
+    def write(self, data):
+        self._input_buf += data
+        msgs = self._input_buf.split("\r")
+        for m in msgs[:-1]:
+            self._parseMessage(m)  # will update _output_buf
+
+        self._input_buf = msgs[-1]
+
+    def read(self, size=1):
+        ret = self._output_buf[:size]
+        self._output_buf = self._output_buf[len(ret):]
+
+        if len(ret) < size:
+            # simulate timeout
+            time.sleep(self.timeout)
+        return ret
+
+    def flush(self):
+        pass
+
+    def flushInput(self):
+        self._output_buf = ""
+
+    def close(self):
+        # using read or write will fail after that
+        del self._output_buf
+        del self._input_buf
+
+    def _addError(self, err):
+        logging.debug("SIM: Adding error %d", err)
+        self._error_stack.append(err)
+
+    def _updateCurrentPosition(self):
+        if not self._moving:
+            self._position = copy.copy(self._target)
+        else:
+            cur_position = {}
+            for axis in range(0, len(self._target)):
+                t = time.time() - self._time_start
+                if self._target[axis] - self._start_pos[axis] > 0:
+                    cur_position[axis] = self._start_pos[axis] + self._speed[axis] * t
```

Comment:
This doesn't really work, because you're not taking into account the accel/decel, which had been taken into account for the time estimation. => It will simulate a move further than requested. Look at the TMCM code, you can just report the position as a ratio of the time passed. (Not exactly simulating the accel yet, but at least it won't go out of bounds). It also avoids having to deal explicitly with negatives.

Reply:
Oh yeah - it is not quite finished - but just a rudimentary step. I was going to add in the acce/decel sim next - then it should be perfect.